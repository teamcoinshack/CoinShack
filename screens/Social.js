import React, { Component } from 'react';
import {
  Text,
  ScrollView,
  View,
  Image,
  StyleSheet,
  FlatList,
  Dimensions,
  TouchableOpacity,
  RefreshControl,
} from 'react-native';
import { Avatar } from 'react-native-elements';
import BellIcon from 'react-native-vector-icons/Entypo';
import { withNavigationFocus } from 'react-navigation';
import Firebase from 'firebase';
import db from '../Database.js';
import Title from '../components/Title.js';
import { background, rowBackground } from '../Masterlist.js';

export default class Social extends Component {

  constructor(props) {
    super(props);

    this.state = {
      friends: [],
      refreshing: true,
    };

    this.refresh = this.refresh.bind(this);
    this.renderRow = this.renderRow.bind(this);
    this.load = this.load.bind(this);
  }

  static navigationOptions = ({ navigation }) => {
    return {
      title: 'Social',
      headerStyle: {
        backgroundColor: background,
        borderBottomWidth: 0,
      },
      headerRight: (
        <TouchableOpacity
          style={{ marginRight: 15 }}
          onPress={() => navigation.navigate('Requests', { 
            callback: navigation.getParam('callback', null),
          })}
        >
          <BellIcon
            name="bell"
            size={24}
            color={navigation.getParam('haveRequests', false)
                   ? '#ff077a'
                   : '#ffffff'}
          />
        </TouchableOpacity>
      ),
      headerTitleStyle: {
        color: '#ffffff',
        fontSize: 20,
      },
      headerTintColor: '#ffffff',
    }
  }

  load(friend) {
    friend.uid === this.state.uid
      ? this.props.navigation.navigate('Profile')
      : this.props.navigation.navigate('FriendsProfile',{
          uid: this.state.uid,
          friendName: friend.username,
          friendUid: friend.uid,
          friendEmail: friend.email,
          callback: this.refresh,
          image: friend.image,
        })
  }
  
  async refresh() {
    try {
      const myUid = Firebase.auth().currentUser.uid;
      let friends = await db.getFriends(myUid); 
      friends = Object.keys(friends);
      friends = await Promise.all(
                    friends.map(async function(uid) {
                      try {
                        let obj = {};
                        const snap = await db.getData(uid);
                        if (!snap) {  
                          db.removeFriend(myUid, uid)
                          return false;
                        }
                        const snapped = snap.val();
                        obj.uid = uid;
                        obj.username = ('username' in snapped)
                                        ? snapped.username
                                        : 'No name :(';
                        obj.email = snapped.email.length > 20 
                                      ? snapped.email.substring(0, 2) + '...'
                                      : snapped.email;
                        obj.value = await db.getTotalValue(uid, snapped);
                        obj.title_id = snapped.title_id;
                        obj.image = await db.getPhoto(uid); 
                        return obj;
                      } catch(error) {
                        console.log(error);
                      }
                    })
                  )
      const myData = await db.getData(myUid);
      const snapped = myData.val();
      friends.push({
        uid: myUid,
        username: snapped.username + ' (You)',
        email: snapped.email,
        value: await db.getTotalValue(myUid, snapped),
        title_id: snapped.title_id,
        image: await db.getPhoto(myUid),
      })
      friends.sort((a, b) => b.value - a.value);
      friends.map((f, index) => {
        f.rank = (index + 1);
        return f;
      });
      let requests = await db.getRequests(myUid);
      requests = Object.keys(requests);
      requests = requests.filter(async function(x) {
        const res = await db.getData(x);
        if (!res) { db.removeFriend(myUid, x) }
        return res;
      })
      this.setState({
        uid: myUid,
        friends: friends,
        refreshing: false,
      })
      this.props.navigation.setParams({ 
        haveRequests: requests.length > 0,
      });
    } catch (error) {
      console.log(error);
    }
  }

  async componentDidMount() {
    try {
      await this.refresh();
      this.props.navigation.setParams({ 
        callback: this.refresh,
      });
    } catch (error) {
      console.log(error);
    }
  }

  renderRow({ item }) {
    if (!item) { return null; }
    const noPic = (
      <Avatar
        rounded
        source={require('../assets/icons/noPic.png')}
        style={styles.imageStyle}
      />
    )
    const havePic = (
      <Avatar
        rounded
        source={{ uri: `data:image/jpg;base64,${item.image}` }}
        style={styles.imageStyle}
      />
    )
    return ( 
      <TouchableOpacity
        onPress={() => this.load(item)}
        style={styles.row}
      >
        {item.rank < 4
          ? ( <Text style={styles['rank' + item.rank]}>{item.rank}</Text> )
          : ( <Text style={styles.rank}>{item.rank}</Text> )}
          {item.image
            ? havePic
            : noPic}
          <View style={{
            flexDirection: 'column',
            alignItems: 'flex-start',
            marginLeft: 10,
          }}>
            <Text style={styles.text1}>{item.username}</Text>
            <Title title_id={item.title_id} />
          </View>
          <View style={{ alignItems: 'flex-end', flex: 1}}>
            <Text style={styles.text2}>
              {'$' + db.stringify(item.value.toFixed(2))}
            </Text>
          </View>
      </TouchableOpacity>
    );
  }
  
  render() {
    const loading = (
      <View style={styles.loading1}>
          <MyBar
            height={65}
            width={Math.round(Dimensions.get('window').width * 0.7)}
            flexStart={true}
          />
      </View>
    );

    const friendsList = (
      <FlatList
        style={styles.flatStyle}
        data={this.state.friends}
        renderItem={this.renderRow}
        keyExtractor={item => item.uid}
        refreshControl={
          <RefreshControl
            refreshing={this.state.refreshing}
            onRefresh={this.refresh}
          />
        }
      />
    );

    const noFriends = (
      <View style={{
        flexDirection: 'row',
        flex: 1,
        justifyContent: 'center',
        marginTop: 30,
      }}>
        <Text style={{
          color: '#7c7c7c',
          fontWeight: '500',
          fontSize: 23,
        }}>
          No friends :(
        </Text>
      </View>
    );

    return (
      <ScrollView 
        style={styles.container}
        contentContainerStyle={styles.contentContainer}
        refreshControl={
          <RefreshControl
            refreshing={this.state.refreshing}
            onRefresh={this.refresh}
          />
        }
      >
          <MyButton
            text="Discover"
            onPress={() => this.props.navigation.navigate('Search', {
              callback: this.props.navigation.getParam('callback', null),  
            })}
            width={Math.round(Dimensions.get('window').width)}
            icon={"magnify"}
          />
          <View style={styles.friendsHeader}>
            <Text style={styles.friends}>Friends</Text>
          </View>
          {this.state.refreshing
            ? loading
            : this.state.friends.length === 0
              ? noFriends
              : friendsList}
      </ScrollView>
    )
  }
}
const styles = StyleSheet.create({
  container: {
    flexDirection: 'column',
    flex: 1,
    alignSelf: "stretch",
    backgroundColor: background,
  },
  contentContainer: {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'center',
    backgroundColor: background,
  },
  friendsHeader: {
    flexDirection: 'row',
    justifyContent: 'flex-start',
    width: Math.round(Dimensions.get('window').width) - 40, 
    marginTop: 10,
  },
  friends: {
    fontSize: 30,
    color: '#dbdbdb',
    fontWeight: '600',
  },
  row: {
    elevation: 1,
    borderRadius: 5,
    backgroundColor: rowBackground, 
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
    width: Math.round(Dimensions.get('window').width) - 28, 
    paddingTop: 15,
    paddingBottom: 15,
    paddingLeft: 10,
    paddingRight: 16,
    marginLeft: 14,
    marginRight: 14,
    marginTop: 0,
    marginBottom: 6,
  },
  imageStyle: {
    width: 50,
    height: 50,
  },
  text1: {
    fontSize: 17,
    color: '#ffffff',
    fontWeight: 'bold', 
  },
  text2: {
    fontSize: 16,
    color: '#dbdbdb',
    fontWeight: 'bold',
  },
  flatStyle: {
    marginTop: 20,
  },
  rank1: {
    color: '#ffd800',
    fontSize: 18,
    fontWeight: 'bold',
    marginRight: 10,
  },
  rank2: {
    color: '#ffffff',
    fontSize: 18,
    fontWeight: 'bold',
    marginRight: 10,
  },
  rank3: {
    color: '#cd7f32',
    fontSize: 18,
    fontWeight: 'bold',
    marginRight: 10,
  },
  rank: {
    color: '#c0c0c0',
    fontSize: 18,
    fontWeight: 'bold',
    marginRight: 10,
  }
});
