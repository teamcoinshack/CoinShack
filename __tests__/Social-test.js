/**
 * @format
 */

import 'react-native';
import React from 'react';
import Social from '../screens/Social.js';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

jest.mock('react-native-google-signin', () => {});
test('renders correctly', () => {
  const tree = renderer.create(<Social />).toJSON();
  expect(tree).toMatchSnapshot();
});
