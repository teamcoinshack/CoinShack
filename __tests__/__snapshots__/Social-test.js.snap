// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "alignItems": "center",
      "backgroundColor": "#373b48",
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={true}
    />
  }
  style={
    Object {
      "alignSelf": "stretch",
      "backgroundColor": "#373b48",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <RCTRefreshControl />
  <View>
    <View
      style={
        Object {
          "height": 65,
          "width": 750,
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#515360",
            "borderRadius": 5,
            "elevation": 1,
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "marginBottom": 10,
            "marginLeft": 14,
            "marginRight": 14,
            "marginTop": 10,
            "opacity": 1,
            "paddingBottom": 10,
            "paddingTop": 10,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "flex-start",
            }
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            }
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#ffffff",
                  "fontSize": 24,
                },
                Object {
                  "paddingRight": 7,
                },
                Object {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            Ôçâ
          </Text>
          <Text
            style={
              Object {
                "color": "#00dce8",
                "fontSize": 20,
                "fontWeight": "700",
              }
            }
          >
            Discover
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginTop": 10,
          "width": 710,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#dbdbdb",
            "fontSize": 30,
            "fontWeight": "600",
          }
        }
      >
        Friends
      </Text>
    </View>
    <View>
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "height": 65,
            "justifyContent": "center",
            "width": 525,
          }
        }
      >
        <View
          animated={true}
          animationConfig={
            Object {
              "bounciness": 0,
            }
          }
          animationType="spring"
          indeterminate={true}
          indeterminateAnimationDuration={1000}
          onLayout={[Function]}
          progress={0.3}
          style={
            Array [
              Object {
                "backgroundColor": undefined,
                "borderColor": "#80e7a5",
                "borderRadius": 4,
                "borderWidth": 0,
                "overflow": "hidden",
                "width": 250,
              },
              undefined,
            ]
          }
          useNativeDriver={false}
        >
          <View
            style={
              Object {
                "backgroundColor": "#80e7a5",
                "height": 6,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateX": -87.5,
                  },
                  Object {
                    "scaleX": 0.30007,
                  },
                ],
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
